sidecar/src/interface.rs

```diff
@@ -317,15 +373,199 @@ impl EnqueuedData {
+            if time + Duration::new(50, 0) > Instant::now() {
+                break;
+            }
+            configs
+                .writers
+                .remove(&configs.last_used.remove(&time).unwrap());
+        }
+    }
+
+    fn start_trace_flusher(self: Arc<Self>, mut force_flush: ManualFuture<()>) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            loop {
+                select! {
+                    _ = tokio::time::sleep(time::Duration::from_millis(
+                        self.interval.load(Ordering::Relaxed),
+                    )) => {},
+                    _ = force_flush => {},
+                }
+
+                let (new_force_flush, completer) = ManualFuture::new();
+                force_flush = new_force_flush;
+
+                let trace_buffer = std::mem::replace(
+                    &mut self.inner.lock().unwrap().traces,
+                    TraceSendData {
+                        send_data: vec![],
+                        send_data_size: 0,
+                        force_flush: Some(completer),
+                    },
+                )
+                .send_data;
+                let mut futures: Vec<_> = Vec::new();
+                let mut intake_target: Vec<_> = Vec::new();
+                for send_data in trace_utils::coalesce_send_data(trace_buffer).into_iter() {
+                    intake_target.push(send_data.target.clone());
+                    futures.push(send_data.send());
+                }
+                for (endpoint, response) in zip(intake_target, join_all(futures).await) {
+                    match response {
+                        Ok(response) => {
+                            if endpoint.api_key.is_none() {
+                                // not when intake
+                                match hyper::body::to_bytes(response.into_body()).await {
+                                    Ok(body_bytes) => self.write_remote_configs(
+                                        endpoint.clone(),
+                                        body_bytes.to_vec(),
+                                    ),
+                                    Err(e) => error!("Error receiving agent configuration: {e:?}"),
+                                }
+                            }
+                            info!("Successfully flushed traces to {}", endpoint.url);
+                        }
+                        Err(e) => {
+                            error!("Error sending trace: {e:?}");
+                            if endpoint.api_key.is_some() {
+                                // TODO: Retries when sending to intake
+                            }
+                        }
+                    }
+                }
+
+                let mut data = self.inner.lock().unwrap();
+                let data = data.deref_mut();
+                if data.traces.send_data.is_empty() {
+                    data.flusher = None;
+                    break;
+                }
```

Comment:
If there are no traces, the flusher will exit, right. But enqueue() (i.e. next received trace) will start a new flusher if there's none.

Reply:
But why not just spawn it once and never exit?

Reply:
To avoid flushers being left-over from connections long closed.
So, it would require adding a synchronization point to notify about new data and adding a notification about the connection being closed. ... But that notification.

The current design was most natural for me to just have the flusher active when needed, then I don't need to worry about synchronizing with it manually.